### PR TITLE
Mesh-Graph Network Example

### PR DESCRIPTION
**Pull request by szaman19**
A partially completed data parallel implementation of `Mesh Graph Network` 

Partially completed as the [Normalizer](https://github.com/CCSI-Toolset/MGN/blob/9c7b1434fefca5f5da233b2c4099a9c54f205848/GNN/utils/Normalizer.py#L6) operation in the original is currently unsupported, but potentially could be simply added via an external layer. It is similar to Batch Norm but with some slight changes. 

Pinging: @bvanessen @tbennun @ndryden

### CODE REVIEW

**Comment by szaman19** [comment]
To do: 

- [ ] Add normalizer of initial input
- [ ] Add an integration test with synthetic data

**Comment by tbennun** [comment]
@szaman19 for normalizer you can use batchnorm

File: applications/graph/MeshGraphNet/Trainer.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import lbann.contrib.args
+import argparse
+import configparser
+import os.path as osp
+from GNN import LBANN_GNN_Model
+
+data_dir = osp.dirname(osp.realpath(__file__))
+
+
+desc = ("Training a Mesh Graph Neural Network Model Using LBANN")
```
Comment (tbennun) [inline applications/graph/MeshGraphNet/Trainer.py:12]:
```suggestion
desc = "Training a Mesh Graph Neural Network Model Using LBANN"
```

File: applications/graph/MeshGraphNet/Trainer.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+# Some training parameters
+
+MINI_BATCH_SIZE = args.mini_batch_size
+NUM_EPOCHS = args.num_epochs
+JOB_NAME = args.job_name
+
+# Some synthetic attributes to get the model running
+DATA_CONFIG = configparser.ConfigParser()
+DATA_CONFIG.read("data_config.ini")
```
Comment (tbennun) [inline applications/graph/MeshGraphNet/Trainer.py:42]: What if this file doesn't exist?

File: applications/graph/MeshGraphNet/Trainer.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    _reader.python.module = classname
+    _reader.python.module_dir = data_dir
+    _reader.python.sample_function = sample
+    _reader.python.num_samples_function = num_samples
+    _reader.python.sample_dims_function = sample_dims
+    return reader
+
+def main():
+  # Use the defaults for the other parameters
+  model = LBANN_GNN_Model(num_nodes=NUM_NODES,
```
Comment (tbennun) [inline applications/graph/MeshGraphNet/Trainer.py:71]: Please run `yapf` on all files.

File: applications/graph/MeshGraphNet/Trainer.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+MINI_BATCH_SIZE = args.mini_batch_size
+NUM_EPOCHS = args.num_epochs
+JOB_NAME = args.job_name
+
+# Some synthetic attributes to get the model running
+DATA_CONFIG = configparser.ConfigParser()
+DATA_CONFIG.read("data_config.ini")
+
+NUM_NODES = int(DATA_CONFIG['DEFAULT']['NUM_NODES'])
```
Comment (tbennun) [inline applications/graph/MeshGraphNet/Trainer.py:44]: How does this work with a dataset that has more than one graph? Is it meant to?

File: applications/graph/MeshGraphNet/GNN.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  split_indices.append(start_index)
+  split_indices.append(split_indices[-1] + node_feature_size)
+  split_indices.append(split_indices[-1] + edge_feature_size)
+  split_indices.append(split_indices[-1] + num_edges)
+  split_indices.append(split_indices[-1] + num_edges)
+  split_indices.append(split_indices[-1] + out_feature_size)
+
+  sliced_input = lbann.Slice(input_layer, axis=0, slice_points=split_indices)
+
+  node_features = lbann.Reshape(lbann.Identity(sliced_input),
```
Comment (tbennun) [inline applications/graph/MeshGraphNet/GNN.py:30]: Why are there `Identity` layers here? Wouldn't direct uses of `sliced_input` work?

File: applications/graph/MeshGraphNet/GNNComponents.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+  def __init__(self,
+               in_dim,
+               out_dim,
+               hidden_dim,
+               hidden_layers,
+               norm_type=lbann.LayerNorm,
+               name=None):
+
+    super().__init__()
+    MLP.global_count += 1
```
Comment (tbennun) [inline applications/graph/MeshGraphNet/GNNComponents.py:21]: do we have to give names to MLPs?

File: applications/graph/MeshGraphNet/GNNComponents.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+  def forward(self, x):
+    """
+      Args:
+        x (Layer) : Expected shape (Batch, N, self.in_dim) 
+
+      Returns: 
+        (Layer): Expected shape (Batch, N, self.out_dim) 
+    """
+    self.instance += 1
+    name = f"{self.name}_instance_{self.instance}"
```
Comment (tbennun) [inline applications/graph/MeshGraphNet/GNNComponents.py:65]: Unused

File: applications/graph/MeshGraphNet/GNNComponents.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+    if self.norm_type:
+      return self.norm_type(x)
+    return x
+
+
+class EdgeProcessor(Module):
+  """ Applies MLP transform on concatenated node and edge features
+  """
+  global_count = 0
```
Comment (tbennun) [inline applications/graph/MeshGraphNet/GNNComponents.py:78]: Same

File: applications/graph/MeshGraphNet/GNN.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+  node_encoder  = MLP(in_dim=in_dim_node, out_dim=out_dim_node,
+                      hidden_dim=hidden_dim_node, hidden_layers=hidden_layers_node,
+                      norm_type=norm_type, name="graph_input_node_encoder")
+
+  edge_encoder  = MLP(in_dim=in_dim_edge, out_dim=out_dim_edge,
+                      hidden_dim=hidden_dim_edge, hidden_layers=hidden_layers_edge,
+                      norm_type=norm_type, name="graph_input_edge_encoder")
+
+  graph_processor = GraphProcessor(num_nodes=num_nodes,
```
Comment (tbennun) [inline applications/graph/MeshGraphNet/GNN.py:66]: Maybe it's worth noting in the code that there is no distinction between mesh edges and world edges in this implementation, nor adaptive remeshing.

File: applications/graph/MeshGraphNet/GNNComponents.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+        source_node_indices (Layer) : Expected shape (Batch, num_edges) 
+        target_node_indices (Layer) : Expected shape (Batch, num_edges)
+      Returns: 
+        (Layer, Layer): Expected shape (Batch, Num_nodes, self.in_dim_node) and 
+          (Batch, num_edges, self.in_dim_edge)
+    """
+
+    for node_processor, edge_processor in self.blocks:
+      x = node_processor(node_features, edge_features, target_node_indices)
+      e = edge_processor(node_features, edge_features,
```
Comment (tbennun) [inline applications/graph/MeshGraphNet/GNNComponents.py:233]: Shouldn't the node processor work on the result of the processed edges at each step? (see Eq. 1 of the paper)

**Review by tbennun — CHANGES_REQUESTED** [review]
Great work! Some comments/questions remain on the model and the files need some standard formatting.

File: applications/graph/MeshGraphNet/GNNComponents.py (hunk 10 above)
Comment (szaman19) [reply applications/graph/MeshGraphNet/GNNComponents.py:233]: Yup, you are right.  `node_processor` and `edge_processor` should be switched. The updated edge features are used for updating node features.

File: applications/graph/MeshGraphNet/SyntheticData.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
+import numpy as np 
+import configparser
+
+
+DATA_CONFIG = configparser.ConfigParser()
+DATA_CONFIG.read("data_config.ini")
```
[the remark this answers is not in the document]
Comment (szaman19) [reply applications/graph/MeshGraphNet/SyntheticData.py:6]: Ended up including an example file in the example.

File: applications/graph/MeshGraphNet/GNNComponents.py (hunk 6 above)
Comment (szaman19) [reply applications/graph/MeshGraphNet/GNNComponents.py]: I usually add it for compute graph visualization and for better debugging messages when things break

File: applications/graph/MeshGraphNet/GNN.py (hunk 9 above)
Comment (szaman19) [reply applications/graph/MeshGraphNet/GNN.py]: Added some documentation in the code and in the read me
